### DISCUSSION TITLE
First steps - example file doesn't work -  no attribute 'interface'

### DISCUSSION THREAD

**Discussion by spaceKelan** [canopen-python/canopen#382] [Q&A] [answered]
Hi all, I hope someone can help me to get this code running.

**Outline**
I want to make my Raspberry Pi talk as CANopen device (NMT Master) via CAN to another device. 
Right now I am working on the Pi side to let it babble and read that data via a Logic Analyzer, as I wait for the hardware to connect the Pi with the other device.

I have a Raspberry Pi 4 and I use a [RS485_CAN_HAT](https://www.waveshare.com/rs485-can-hat.htm) from waveshare with an MCP2515 IC. The Hat documentation is very basic.

When I use a basic CAN script to send data via CAN or when I use `cansend can0 123#42ff` I receive the data on my PC via the Logic Analyzer just fine. ( i.e. I presume my `/boot/config.txt` setup is fine and I configured the vcan can0 properly, the baudrates are the same)
![image](https://github.com/christiansandberg/canopen/assets/51329415/dfa307bf-89c2-4228-b35f-13919ec6c431)

However, as I want to use CANopen and make use of the application layer functionality I tried the script provided in the README and also the example. However, in both cases I perceive the error
```
    network.connect(bustype='socketcan',channel='can0',bitrate=100000)
  File "/home/pi/Codes/ma/canopen/canopen/network.py", line 113, in connect
    self.bus = can.interface.Bus(*args, **kwargs)
AttributeError: 'NoneType' object has no attribute 'interface'
```
I tried to cut down the provided `eds`-File, as I thought it might be some configuration issue, but this did not resolve the problem.

From reading CANopen 301 I did not get the impression, that there is some crucia device-specific setup that would make the network.connect() function fail due to incorrect values. Maybe I am wrong.
Unfortunately all EDS are quite overwhelming by their size so it contains plenty of potential error sources but as the default eds provided does not work either, I would really appreciate any help!

Also, I tried to use @bggardner  `canopen-rpi` and got stuck at the same position as I always receive the `BUS_OFF error`. 
![image](https://github.com/christiansandberg/canopen/assets/51329415/6ffae23a-9318-42ef-b4fb-1c342b19da71)

So in both cases I try to talk via a bus, which is not yet configured. 
Clearly I see that I am missing something obvious.

I have some questions:
1. Do I need to consider the specific board for basic functionality in the EDS and if yes, how? 
2. What am I doing wrong so that the bus is not running? How can I resolve that issue? I only set `BaudRate_1000=1` flag as this is the baudrate I set in the python script as well
3. Is the problem, that I am trying to use a bus but only 1 device (Pi) is attached?
I assumed that as long as the Pi is the master, this should not be the cause.
4. How do I tell in the EDS that the Pi is the (NMT) master? I presumed it is the `SimpleBootUpMaster=1` flag.
5. Does it have to do with the PRE-OPERATIONAL state and if so, how can I check? 

I am looking forward to your reply, please don't hesitate to ask for further details for debugging.

Thank you in advance!

spaceKelan

**Comment by bggardner** [comment]
I can't speak to `christiansandberg/canopen`'s implementation, but be aware that `bggardner/canopen-rpi` uses a [fork](https://github.com/bggardner/python-can) of `python-can` that implements the `BusState` class a little differently - see the full discussion [here](https://github.com/hardbyte/python-can/issues/736).  This may be the cause of your errors with `bggardner/canopen-rpi` and the `socketcanopen` module.  I've been using `socketcanopen` to run a CANopen master for a few years now, so hopefully I can help.

First, can you run the [python-can example](https://github.com/hardbyte/python-can/tree/develop#example-usage) (using `can0` instead of `vcan0`)?

1. I'm not sure what you mean by this, but `canopen-rpi` relies on the EDS to determine if the node performs varies CANopen functions.
2. The CAN bitrate is set when you bring up the `can0` interface on the Pi (`sudo ip link set can0 up type can bitrate 1000000`), not in Python.  If `cansend` works, then the problem is not related to the bitrate, and it should not be passed as an argument with `bustype='socketcan'`.
3. CAN requires two nodes on the bus for data frames to be transmitted successfully, otherwise the bus will be loaded with error frames.  The Raspberry Pi is one node, and  your Logic Analyzer may be acting as a second node.  You can troubleshoot this by running `ip -d link show can0`.  If the `state` is `ERROR-ACTIVE`, that's good.  Anything else means there's something wrong.  Even if it initially comes up as `ERROR-ACTIVE`, trying to send a frame with `cansend` may cause it to switch to `ERROR-PASSIVE`, which probably means there's no other nodes on the bus.
4. If bit 0 of object 0x1F80 is set, the device is considered an NMT Master (defined in CiA 302-2 v4.1.0).
5. I don't believe the NMT state is related to your problem, at least at this point.

**Reply by spaceKelan** [reply]
First of all, **thank you so much for the quick detailed response**!

a) I checked for the discussion linked, I hope I got it right: In the `canopen-rpi` repository the `can.BusABC.states` align with ISO 11898-1 6.13 Error Active, 6.14 Error Passive and 6.15 Bus-off (node is in state bus off, if it got sent there by FCE request) and in Table 2 (7.4) it states Node_status indicates if the node is in bus-off state. Therefore, **bus-off clearly refers to the node state** not the bus state.
Anyways, how can I recover my RaspberryPi-node from `BUS_OFF` to `ERROR_ACTIVE`?

b) I went to your `canopen-rpi ` repository and created an `example_python_can.py` file containing the above mentioned example with `can0`. I executed `sudo modprobe vcan` and `sudo ip link set can0 up type can bitrate 1000000` to run it at 1 MHz. 
I forgot to mention: as I had issues with python dependencies I run a virtual environment in your repository 
When I execute `python3 example_python_can.py` I receive the following output in the **Pi's terminal**
![image](https://github.com/christiansandberg/canopen/assets/51329415/09aec866-847c-46a4-a6c6-83a8cde445d0)
while the **Logic Adapter** reads the CAN bus:
![image](https://github.com/christiansandberg/canopen/assets/51329415/54febbb6-a6b8-4971-83d7-6f9dfb65f102)
I suppose this is good, I do not understand why the bus generates so many can_errors, I suppose it is due to the missing `ACK` from another device.
_c) Also - It seems to be the first time that I receive packets on CAN High (top row, green) and CAN Low (bottom row, blue) when running `socketcan` - I still wonder what's going on with the differential signal. ![image](https://github.com/christiansandberg/canopen/assets/51329415/107f1fb0-dfe6-416b-9ab7-f2cb4054476a)
Nevertheless, other than expected the data does not coallign but "mirrored", instead CAN Low (blue) displays ID-Field and control field before can_errors come in. I assume that the message on blue is the RTR and green is the response_
(Even when I worked with CAN only, I only ever got messages on one line.
![image](https://github.com/christiansandberg/canopen/assets/51329415/f4bc335f-2be8-43dd-bd62-5b4d288e8987))
Why is data sent differently on CAN H and L and not in a mirrored fashion?
What do you learn from this? Please let me know what information I can provide for debugging.

1. I mean - the EDS is configuring the use of the CANopen 301 standard. It has all kind of 301-specific flags and definitions. Even though it provides the use of manufacturer-specific fields, they are not "mission-critical". Therefore, I suppose my EDS needs to fullfil the basic functionality (like the minimal NMT master example `node.eds` provided in `bggardner/canopen-rpi/examples`) I could just copy that eds file into another framework and use it there, as long as the script I execute it with calls the existing OD indexes and subindexes.
2.  If I understood you correctly, I set all flags to`BaudRate_XXX=0` but still I receive the `BUS_OFF` state. Setting all to `=1` results in the same node state.
3. Shoot, I expected this as no ACK entity is present:
![image](https://github.com/christiansandberg/canopen/assets/51329415/11ced06a-fc6b-4e6b-9923-5bad626f13cc)
The `ERROR-PASSIVE`-state explains all the `can_error` messages. (6.14 a node in the error-passive state will send the passive error flag consisting of 6 consecutive recessive bits and waits for a while before retransmitting the initial packet)
While writing my hardware arrived - so I hope this resolves within the next days
The difference when working not with CANopen but CAN the other day was, that also with CAN I connected some Arduinos as nodes but I did not look into Arduino and CANopen yet, as I wanted to not get completely lost in these "dummy devices". I think I will give it a try with 
[`curvi/CANopen`](https://github.com/curvi/CANopen) as a first step.
4. This explains why I didn't know about this as all `302` is only available for members. I assumed I try if  `301` on top of `ISO 11898-1/2` is sufficient. (Sounds a little naive but I thought it is worth a shot and it is my first project working with databusses and OSI in such detail - also there is an ECSS standard which describes the setup in between 301 and ISO-11898-1/2)
As `302`  is not in the extensive ``CANopenNode` I assumed this is negligible. 
5. Thank you!

Sincerely,

spaceKelan

**Comment by bggardner** [comment, marked as answer]
a) The best way to recover from `BUS_OFF` is to restart the `can0` interface, especially at this point in debugging.  It should eventually recover automatically if there is non-error traffic on the bus when its internal error counters fall below thresholds.

b/c) The output of `example_python_can.py` doesn't seem right, probably due to the error frames on the bus.  You shouldn't use the logic analyzer to look at CANH/CANL individually, but after a transceiver, like at the CAN_RXD signal on your [CAN HAT schematic](https://www.waveshare.com/w/upload/1/1d/RS485_CAN_HAT_Schematic.pdf).  There should always be another CAN node on the bus, as it will set the ACK bit when another node is transmitting a frame, preventing error frames.  Note that a CAN node means a CAN controller (like a MCP2515), not just a transceiver.  Also, make sure your physical CANH/CANL bus wiring is properly terminated with 120 ohms at each end of the bus.  The cabling should be a twisted pair (or 3 with the ground) and have a characteristic impedance of 120ohms, but for short lengths (<1m), this is less critical.

1. The example `node.eds` is for a slave node, not a CANopen NMT master.  Use caution when referring to a CANopen master, as there are many functions that may make up what a typical master performs ("NMT master" sends NMT/Node Control messages, "SYNC producer sends SYNC messages", etc.), but any node could be configured to perform these functions (especially if they are flying masters).
2. The `BaudRate` entry in the EDS does not affect the physical bitrate for a `socketcan` interface.  It is ignored.  To repeat, the bitrate for a `socketcan` interface is not set by the `BaudRate` entry in the EDS.
3. See (b/c) above.
4. Here's an older, publicly-available version of [CiA 302](http://overpof.free.fr/schneider/CAN%20&%20CANopen/CANopen/%A9CiA%20CANCANopen%20CD%20V5.1/standard/dsp30232.pdf), which is probably all you need to get by.
5. Kein Problem!

**Reply by spaceKelan** [reply]
Sorry for the late reply, I didn't manage to find the time on Friday.

a)/b)/c) Thank you, I wanted to implement the new device today/tomorrow and let you know if it worked. I'll move the logic analyzer to receive the pure `CAN_RXD` signal - I thought looking at it separately would make sense as it would display the mirrored signals. Why is this assumption wrong/Why should it always looked at after the CAN controller?

1. In the network I want to create I want to have a single NMT master. 
2. Copy!

4. I am looking into the CiA 302 right now to grasp if I require that functionality apart from the 0x1F80 to make it run. No further questions right now.

**Comment by bggardner** [comment]
It's not wrong to look at the differential signals, but not ideal.  When you connect the scope probes, it creates a long "stub", which alters the topology/impedance of the bus, which can lead to reflections and bit errors.  Ideally all transceivers are connected directly (with a stub as short as possible) to the bus.  CAN likes an end-to-end "bus" topology, not radial, star, mesh, or otherwise.

**Reply by spaceKelan** [reply]
thank you!

**Comment by ericbaril72** [comment]
I suggest you simplify until you reach milestones…

first, just send one raw message without CANopen library… just bare CAN.
- your logic analyser seems like Logic from saleaé. The decoder need
threshold voltages and this would easily explain your can-h/l not being
mirrored.
- the Can errors ar because there is a single device and no receiving the
‘ACK’ bit from a secondary not… so the hardware can module automatically
retransmits the message.

If yours are using the MCP2515 as can controller, there is a possible
setting to ‘send-once’ but would only do that if I were you to validate
that you know how to change this setting.  When multiple nodes on the bus,
I do not know to which extent does the mcp2515 would repeat in case the
message sent is of lower priority.

Once you can reliably send and receive a message ( get a second node on the
bus to help you out)!you can go to step 2-> start messing with CANopen or
j1939

Otherwise, you are questioning the CANopen stack when the “interface” with
the BUS class is what your are struggling with.

Get any second CAN node on the bus and use you logic analyser to figure out
it’s transmit baud and try configuring the PI for RX and check if received
message matches your logic analyser.

Once your baud rate is ok, you can transmit anything and the other node
will “ACK” even if it doesn’t mean anything (j1939 vs CO)

Goodluck

**Reply by spaceKelan** [reply]
Hi @ericbaril72,

thank you for your response. 

1. Yes my logic analyser is from salea, may you please elaborate on the threshold explains it?
2. Two nodes are the minimum, figured that out by now, too.
3. I did add the second device after receival of that CAN interface and CAN 2.0B is supported and works well.

I'm sorry if this is the wrong space to ask this:
I plan to move from 

| Device Name | Computer| OS|
| ----------- | ----------- |----------- |
| A| Pi| Raspbian |
| B| Desktop |older Fedora Linux        |

to other hardware. The pi is therefore just for testing purposes. I am currently organising myself a second pi so I can use this repository here without interfacing with other frameworks focussing on OD and simple implementation of it like this


| Device Name | Computer| OS|
| ----------- | ----------- |----------- |
| A| Pi| Raspbian |
| B| Pi| Raspbian |


My target implementation looks like this though:

| Device Name | Computer| OS|
| ----------- | ----------- |----------- |
| A| something exotic | RTEMS|
| B| Desktop |older Fedora Linux        |

When I tried to make `CANopen` run with a self-created minimum feasible OD, I also looked into `CANopenLinux` for `device B`, which turns out to create issues in the `309 ASCII Gateway`. 

Next I used their `template OD.eds` and generated `OD.h` and `OD.c` files, which did not built the code due to errors! After adding elements to a similar level as my personal minimum feasible OD, I could built but would get the same error when executing.

I know this is a different repository, I stalled working with this for now, as it seems too complex to strip, as I won't need `309` in my final implementation anyways. However, I am even more confused now.

4. Is it possible to take two CAN devices and implement CANopen protocol on them even though they did not offer that functionality before? ( I assume yes, but `301` reads like there would have to be a `Vendor ID` noted at CiA for this to work, which I doubt.)

My current plan is to implement 
+ SDO block upload and download for example frames 
+ some PDOs and go from there.

The only CANopen ICD I was able to get does not go into detail regarding NMT etc. , but only contains TPDO and RPDO. Therefore the whole CANopen complexity is still a bit confusing.
 
The implementation of NMT seems complex and unfortunately since this repository is based on python and my target language must be C++, I can only learn from this, but not implement this directly. However, I will definitely check how I can inspire myself by this work as this seems way more straight-forward than `CANopenLinux`.

Looking forward to hear your thoughts! 

Thank you for your time!

**Comment by ericbaril72** [comment]
@Andy,

Thresholds: voltage level at which the logic analyser will report a 0 or 1.

CAN low will be either 2.5V for “1” or idle, and ~1.5v for 0’s.   Which is
still considered a ‘1’ in TTL/CMOS logic.
If using an analog  input , check if you can set the threshold voltage
where the ‘decoder’ will return a 0 or 1.

Most likely not avail on digital only inputs devices.

4.Vendor ID —> yes if you plan on selling a commercial product. Otherwise,
make something up

====
I did the same as you do…. Using the Python library to learn and practice
and validate my “c” version running on embedded CPU and even if for Codesys
St code where I did not like the limitations of codesys in regards to their
parameter manager.

Eric

**Reply by spaceKelan** [reply]
thank you very much. I'll let you know how it goes, I need some days as I need to get my second pi back and all.
